[Activejob] Enable enqueue_after_transaction_commit by default with ActiveRecord and validate invalid configurations

mugitti9 opened the pull request:
### Motivation / Background
Until now, the default value of ```enqueue_after_transaction_commit``` has been false.

This meant that jobs enqueued inside a transaction were dispatched to the queue immediately and could still run even if the transaction was rolled back. In practice, this is not a rare edge case—jobs that depend on uncommitted data often end up running prematurely, breaking application consistency.

With this change, the default is switched to true only when ActiveRecord is defined. This ensures that jobs are enqueued only after the surrounding transaction has successfully committed, preventing jobs tied to rolled-back work from being executed unnecessarily.


As DHH also stated in [#53375](https://github.com/rails/rails/pull/53375#issuecomment-2555694252)
```Yes, by default, all jobs should enqueue after the commit.``` 
I strongly agree with this position. At the same time, this change does not remove flexibility: if immediate enqueueing is still desired, it can be explicitly enabled by setting ```self.enqueue_after_transaction_commit = false``` on a per-job basis.

### Detail

- The default value of ```enqueue_after_transaction_commit``` is now set to true only when ActiveRecord is available.
- A validation has been added: attempting to set ```enqueue_after_transaction_commit = true``` when ActiveRecord is not defined will now raise an error, since such a configuration would not have meaningful semantics without transaction hooks.

### Why this change is needed (what problem it solves)
With the current behavior on main, jobs can be enqueued before the surrounding database transaction has successfully committed. For example:

```rb
ActiveRecord::Base.transaction do
  user = User.create!(name: "James")
  UserCreationNotifyJob.perform_later(user:)
end
```

In this case, UserCreationNotifyJob may be enqueued while the user record is still uncommitted. If the transaction is rolled back, the job can still run and attempt to operate on data that does not actually exist. This leads to subtle and easily overlooked inconsistencies.

Until Rails 7.2, most adapters—Sidekiq, Resque, and Solid Queue—effectively behaved as if enqueue_after_transaction_commit = true by default. As a result, many applications implicitly relied on jobs being enqueued only after the transaction had fully committed. With the current default on main, however, jobs may run earlier than expected, and this can introduce hard-to-debug issues.

In our application, we currently mitigate this by explicitly setting enqueue_after_transaction_commit = true on almost all jobs. But this shifts responsibility for a fundamental safety guarantee from the framework to individual applications, increases boilerplate, and risks inconsistencies when someone forgets to set it.

This issue is not theoretical or rare. For example, at a recent Ruby/Rails conference in Japan ("Kaigi on Rails"), there was a well-received talk titled “Never enqueue async jobs inside a transaction—seriously, never!”, which highlighted exactly this class of problems. The talk resonated widely, which shows how many developers have been impacted by premature job execution.

### Additional information

When ActiveRecord is not defined, transaction callbacks are not available, and therefore enabling ```enqueue_after_transaction_commit``` would result in misleading or inconsistent behavior. Explicit validation has been added to surface this misuse early and clearly.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

jeremy reviewed (commented):
-1 on the AR dep. Would either enqueue immediately if AR is unavailable, gracefully behaving as anyone would expect, or raise an exception indicating that enqueue_after_transaction_commit should be set to false.

mugitti9 committed:
[ActiveJob] Enable enqueue_after_transaction_commit by default when ActiveRecord is available

mugitti9 commented:
@jeremy 
Thanks for your review and for pointing that out!

You're absolutely right — adding ActiveRecord as a hard dependency was too much.
I've updated the implementation based on your suggestion:

- The default value of enqueue_after_transaction_commit is now set to true only when ActiveRecord is defined.
- When ActiveRecord is not available, setting enqueue_after_transaction_commit = true doesn’t make sense, so an exception is raised in that case to surface the misconfiguration clearly.

I'd appreciate it if you could take another look.

mugitti9 committed:
Raise an error when enqueue_after_transaction_commit is true without ActiveRecord

mugitti9 committed:
Fix Railties test expectation for enqueue_after_transaction_commit default

jeremy committed:
ActiveJob: decouple enqueue_after_transaction_commit default

jeremy committed:
Railties: cover enqueue_after_transaction_commit defaults

Also cover explicit false override for new apps.

jeremy committed:
Railties: add 8.2 default for enqueue_after_transaction_commit

Also group Active Job defaults in the 8.2 new framework defaults template.

# Conflicts:
#	railties/lib/rails/application/configuration.rb
#	railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_8_2.rb.tt

jeremy committed:
Docs: note enqueue_after_transaction_commit default

# Conflicts:
#	activejob/CHANGELOG.md

zzak commented:
I'm confused that we would re-introduce this config (see: 582f91e), but somehow makes sense in the context of changing the default? :thinking:

zzak commented:
Unless I'm missing something, the flag itself was deprecated and removed in 8.1, and we'd be re-introducing it in 8.2 / 9.0:
https://github.com/rails/rails/blob/582f91e9330071200b1b72bcfef46a6960ea37e3/activejob/lib/active_job/railtie.rb#L33-L38

jeremy commented:
@zzak, revised to avoid re‑introducing global config—for now. I think it's an awkward mistake, though. You can see how much worse the code is. Not going to foist that upon anyone. Apps need an app-wide toggle for this. I aim to un-deprecate, particularly considering it wasn't removed on schedule.

zzak commented:
Yeah, if you want to change the default, old apps need a way to maintain the existing behavior without any changes so a global config is kind of the only option. :thinking: I was just pointing out that bit of context.

zzak commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_8_2.rb.tt:
```diff
@@ -9,6 +9,19 @@
+#++
+# ActiveSupport.on_load(:active_job) do
+#   ActiveJob::Base.enqueue_after_transaction_commit = true
+# end
```
I think it's a bit of an anti-pattern to put initializer code here, typically we only put `Rails.application.config` afaik

Maybe it's ok to keep the config for this since the previous deprecation cycle is concluded, and we give people at least a MINOR version to adopt the change.

jeremy replied:
Right, this is the ugly code foisted upon apps referenced above. I'm changing it back to global config and un-deprecating.

jeremy committed:
ActiveJob: apply enqueue_after_transaction_commit config

jeremy committed:
Railties: restore enqueue_after_transaction_commit default config

jeremy committed:
Railties: cover enqueue_after_transaction_commit config overrides

jeremy committed:
Docs: restore enqueue_after_transaction_commit config guidance

jeremy reviewed (approved):
Un-deprecated the global config toggle, added docs, switched to versioned config (`load_defaults "8.2"` with the new framework defaults initializer), and gated config on AR load (rather than presence of the constant).
